plugins/upower.c

```diff
@@ -0,0 +1,276 @@
+		ofono_error("%s: iface != TYPE_STRING!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_get_basic(&iter, &iface);
+
+	/* Check for UPower.Device iface */
+	if (g_str_equal(iface, UPOWER_DEVICE_INTERFACE) != TRUE) {
+		ofono_error("%s: wrong iface: %s!", __func__, iface);
+		return FALSE;
+	}
+
+	/* advance iter... */
+	if (!dbus_message_iter_next(&iter)) {
+		ofono_error("%s: advance iter failed!", __func__);
+		return FALSE;
+	}
+
+	/* @nd param is Dict; validate wrapper is an array */
+	if (dbus_message_iter_get_arg_type(&iter) != DBUS_TYPE_ARRAY) {
+		ofono_error("%s: type != ARRAY!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_recurse(&iter, &dict);
+
+	while (dbus_message_iter_get_arg_type(&dict) == DBUS_TYPE_DICT_ENTRY) {
+		DBusMessageIter entry, val;
+		const char *key;
+
+		dbus_message_iter_recurse(&dict, &entry);
+
+		if (dbus_message_iter_get_arg_type(&entry) != DBUS_TYPE_STRING) {
+			ofono_info("%s: key type != STRING!", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&entry, &key);
+
+		if (g_str_equal(key, "Percentage") != TRUE) {
+			ofono_info("%s: key != 'Percentage': %s", __func__, key);
+
+			dbus_message_iter_next(&dict);
+			continue;
+		}
+
+		ofono_info("%s: key == 'Percentage'", __func__);
+
+		dbus_message_iter_next(&entry);
+
+		if (dbus_message_iter_get_arg_type(&entry) != DBUS_TYPE_VARIANT) {
+			ofono_info("%s: 'Percentage' val != VARIANT", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_recurse(&entry, &val);
+
+		if (dbus_message_iter_get_arg_type(&val) != DBUS_TYPE_DOUBLE) {
+			ofono_info("%s: 'Percentage' val != DOUBLE", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&val, &percentage_val);
+		break;
+	}
+
+	if (percentage_val == 0.00) {
+		battery_val = 0;
+	} else if (percentage_val > 0.00 && percentage_val <= 100.00) {
+		battery_val = ((int) percentage_val - 1) / 20 + 1;
+        } else {
+		ofono_error("upower: Invalid value for battery level: %f",
+								percentage_val);
+		return FALSE;
+	}
+
+	DBG("done parsing upower message; last_battery_val: %d "
+							"battery_val: %d",
+							last_battery_val,
+							battery_val);
+	if (last_battery_val == battery_val)
+		return FALSE;
+	else
+		last_battery_val = battery_val;
+
+	for (list = modems; list; list = list->next) {
+		struct ofono_modem *modem = list->data;
+
+		__ofono_modem_foreach_registered_atom(modem,
+					OFONO_ATOM_TYPE_EMULATOR_HFP,
+					emulator_battery_cb,
+					GINT_TO_POINTER(battery_val));
+	}
+
+	return TRUE;
+}
+
+static void emulator_hfp_watch(struct ofono_atom *atom,
+				enum ofono_atom_watch_condition cond,
+				void *data)
+{
+	struct ofono_modem *modem = data;
+
+	if (cond == OFONO_ATOM_WATCH_CONDITION_REGISTERED) {
+		DBG("%s: REGISTERED", __func__);
+
+		modems = g_list_append(modems, modem);
+
+		if (modems->next != NULL)
+			return;
+
+		upower_battery_watch = g_dbus_add_signal_watch(connection,
```

Comment:
Ok, I see that you are registering this only for one modem, also, there is only one that currently can have a HFP atom. Taking a deeper look at the code it also seems that the case of more than one modem with HFP would be handled correctly.

However, the first part of my previous comment is still correct:

1. we should retrieve the current value for the battery right after having registered for the signal
2. we need to set an initial value for the battery level for a modem just added to the list